test/unit/handlebars.cpp

```diff
@@ -0,0 +1,460 @@
+    for (auto [firstname, lastname]: first_and_last_names) {
+        person = {};
+        person.set("firstname", firstname);
+        person.set("lastname", lastname);
+        peopleObj.set(firstname, person);
+    }
+    context.set("peopleobj", peopleObj);
+    context.set("author", true);
+    context.set("firstname", "Yehuda");
+    context.set("lastname", "Katz");
+    dom::Array names = dom::newArray<dom::DefaultArrayImpl>();
+    names.emplace_back("Yehuda Katz");
+    names.emplace_back("Alan Johnson");
+    names.emplace_back("Charles Jolley");
+    context.set("names", names);
+    dom::Object namesobj;
+    namesobj.set("Yehuda", "Yehuda Katz");
+    namesobj.set("Alan", "Alan Johnson");
+    namesobj.set("Charles", "Charles Jolley");
+    context.set("namesobj", namesobj);
+    dom::Object city;
+    city.set("name", "San Francisco");
+    city.set("summary", "San Francisco is the <b>cultural center</b> of <b>Northern California</b>");
+    dom::Object location;
+    location.set("north", "37.73,");
+    location.set("east", "-122.44");
+    city.set("location", location);
+    city.set("population", 883305);
+    context.set("city", city);
+
+    dom::Object lookup_test;
+    dom::Array people_lookup = dom::newArray<dom::DefaultArrayImpl>();
+    people_lookup.emplace_back("Nils");
+    people_lookup.emplace_back("Yehuda");
+    lookup_test.set("people", people_lookup);
+    dom::Array cities_lookup = dom::newArray<dom::DefaultArrayImpl>();
+    cities_lookup.emplace_back("Darmstadt");
+    cities_lookup.emplace_back("San Francisco");
+    lookup_test.set("cities", cities_lookup);
+    context.set("lookup_test", lookup_test);
+
+    dom::Object lookup_test2;
+    dom::Array persons = dom::newArray<dom::DefaultArrayImpl>();
+    dom::Object person1;
+    person1.set("name", "Nils");
+    person1.set("resident-in", "darmstadt");
+    persons.emplace_back(person1);
+    dom::Object person2;
+    person2.set("name", "Yehuda");
+    person2.set("resident-in", "san-francisco");
+    persons.emplace_back(person2);
+    lookup_test2.set("persons", persons);
+    dom::Object cities;
+    dom::Object darmstadt;
+    darmstadt.set("name", "Darmstadt");
+    darmstadt.set("country", "Germany");
+    cities.set("darmstadt", darmstadt);
+    dom::Object san_francisco;
+    san_francisco.set("name", "San Francisco");
+    san_francisco.set("country", "USA");
+    cities.set("san-francisco", san_francisco);
+    lookup_test2.set("cities", cities);
+    context.set("lookup_test2", lookup_test2);
+
+    /////////////////////////////////////////////////////////////////
+    // Register helpers
+    /////////////////////////////////////////////////////////////////
+    Handlebars hbs;
+    helpers::registerAntoraHelpers(hbs);
+    hbs.registerHelper("progress", [](dom::Array const& args) -> dom::Value {
+        if (args.size() < 3) {
+            return fmt::format("progress helper requires 3 arguments: {} provided", args.size());
```

Comment:
How does this error reporting work? Errors are strings? I don't understand this...

Reply:
I haven't worked much on error reporting yet because it's some area where we probably want to deviate from javascript. 

Javascript always throws for any small thing and that's it. And the helpers can throw, so nothing the engine can do about that. Then there are hooks for missing variables and helpers: https://handlebarsjs.com/guide/hooks.html#helpermissing

But helpers can do anything. In this case, the "progress" helper just renders the error instead of throwing it because this is much more useful in tests. Instead of an exception is CI, this becomes just another line where the output was different when we diff.